src/walker.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+import { cloneDeep } from 'lodash';
+
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Pipeline;
+  stepFn?: (step: Step) => Step;
+  commandFn?: (command: Command) => Command;
+}
+
+export async function evaluatePipeline(p: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const pipeline = new Pipeline(p.name);
+  pipeline.steps = [];
+  for (const step of p.steps) {
+    let newStep: PotentialStep | null = cloneDeep(step);
+    newStep = await evaluateStep(newStep, conditionalCache);
+    if (newStep) {
+      pipeline.steps.push(newStep);
+    }
+  }
+  return pipeline;
+}
+
+async function evaluateStep(
```

Comment:
can we reuse `unwrapSteps` on the steps themselves + their dependencies instead?

Reply:
I'm definitely surprised about this being a new method, I thought the serialiser already had to do the equivalent of this method.

Reply:
yeah, will remove `evaluateStep` and call `unwrapSteps` instead

Reply:
You're right, I looked where `unwrapSteps` is used, and there's two places:
1. `sortedSteps` only uses it for the top level steps, it unwraps the dependencies itself.
2. `Step.toJson` uses it to unwrap both dependencies and effects together, but I think that's wrong: that would mean that when they're serialised to json steps don't declare dependencies on conditional steps that aren't accepted (which they should unconditionally) or on conditional steps that are effects that are depended on by other steps in the same pipeline (which they should).

@joscha Can you double check that?